erc1155/src/lib.rs

```diff
@@ -185,56 +187,85 @@ impl ExtendERC1155TokenBase for ERC1155Token {
                 return false;
             }
         }
-
         true
     }
 
-    fn mint(&mut self, account: &ActorId, id: &u128, amount: u128) {
+    fn mint(&mut self, account: &ActorId, id: &u128, amount: u128, meta: Option<TokenMetadata>) {
         if account == &ZERO_ID {
             panic!("ERC1155: Mint to zero address")
         }
+        if let Some(metadata) = meta {
+            // Off you go...
+            if amount > 1 {
+                panic!("ERC1155: Mint metadata to a fungible token")
+            }
+            self.token_metadata.insert(*id, metadata);
+        }
         let prev_balance = self.balance_of(account, id);
         self.set_balance(account, id, prev_balance.saturating_add(amount));
     }
 
-    fn mint_batch(&mut self, account: &ActorId, ids: &[u128], amounts: &[u128]) {
+    fn mint_batch(
+        &mut self,
+        account: &ActorId,
+        ids: &[u128],
+        amounts: &[u128],
+        meta: &[Option<TokenMetadata>],
+    ) {
         if account == &ZERO_ID {
             panic!("ERC1155: Mint to zero address")
         }
 
         if ids.len() != amounts.len() {
             panic!("ERC1155: ids and amounts length mismatch")
         }
+        ids.iter()
+            .enumerate()
+            .for_each(|(idx, x)| self.mint(account, x, amounts[idx], meta[idx].clone()));
+    }
 
-        for (i, ele) in ids.iter().enumerate() {
-            let amount = amounts[i];
-            let prev_balance = self.balance_of(account, ele);
-            self.set_balance(account, ele, prev_balance.saturating_add(amount));
+    fn burn(&mut self, id: &u128, amount: u128) {
+        let owner = &msg::source();
+        if self.can_burn(owner, id, amount) {
+            self.set_balance(
+                &msg::source(),
+                id,
+                self.get_balance(owner, id).saturating_sub(amount),
+            );
         }
     }
 
     fn burn_batch(&mut self, ids: &[u128], amounts: &[u128]) {
-        let owner = &msg::source();
-
         if ids.len() != amounts.len() {
             panic!("ERC1155: ids and amounts length mismatch")
         }
 
-        if !self.owner_of_batch(ids) {
-            panic!("ERC1155: have no ownership of ids")
-        }
+        ids.iter().enumerate().for_each(|(idx, x)| {
+            if !self.can_burn(&msg::source(), x, amounts[idx]) {
+                panic!("ERC1155: all batch element should be burnable");
+            }
+        });
```

Comment:
In my opinion `for_each` doesn't introduce any benefit over `for` when contains multiple lines inside.
```suggestion
        for (id, amount) in ids.iter().zip(amounts) {
            if !self.can_burn(&msg::source(), id, *amount) {
                panic!("ERC1155: all batch element should be burnable");
            }
        }
```